cores/arduino/stm32/LowPower.c

```diff
@@ -0,0 +1,289 @@
+  * @param  None
+  * @retval None
+  */
+void LowPower_init(){
+  /* Enable Power Clock */
+  __HAL_RCC_PWR_CLK_ENABLE();
+
+  /* Ensure that HSI is wake-up system clock */
+  __HAL_RCC_WAKEUPSTOP_CLK_CONFIG(RCC_STOP_WAKEUPCLOCK_HSI);
+
+  /* Check if the system was resumed from StandBy mode */
+  if (__HAL_PWR_GET_FLAG(PWR_FLAG_SB) != RESET)
+  {
+    /* Clear Standby flag */
+    __HAL_PWR_CLEAR_FLAG(PWR_FLAG_SB);
+  }
+
+  /* Clear all related wakeup flags */
+  __HAL_PWR_CLEAR_FLAG(PWR_FLAG_WU);
+}
+
+/**
+  * @brief  Configure a pin as wakeup source if compatible.
+  * @param  pin: pin to configure
+  * @param  mode: pin mode (edge or state). The configuration have to be compatible.
+  * @retval None
+  */
+void LowPower_EnableWakeUpPin(uint32_t pin, uint32_t mode) {
+#ifndef STM32L4xx
+  UNUSED(mode);
+#endif
+  PinName p = digitalPinToPinName(pin);
+  uint32_t wupPin = pinmap_find_function(p, PinMap_WKUP);
+  switch (wupPin) {
+#ifdef PWR_WAKEUP_PIN1
+    case 1 :
+    #ifdef STM32L4xx
+      if (mode == RISING) {
+        HAL_PWR_EnableWakeUpPin(PWR_WAKEUP_PIN1_HIGH);
+      }
+      else {
+        HAL_PWR_EnableWakeUpPin(PWR_WAKEUP_PIN1_LOW);
+      }
+    #else
+      HAL_PWR_EnableWakeUpPin(PWR_WAKEUP_PIN1);
+    #endif
+      break;
+#endif //PWR_WAKEUP_PIN1
+#ifdef PWR_WAKEUP_PIN2
+    case 2 :
+    #ifdef STM32L4xx
+      if (mode == RISING) {
+        HAL_PWR_EnableWakeUpPin(PWR_WAKEUP_PIN2_HIGH);
+      }
+      else {
+        HAL_PWR_EnableWakeUpPin(PWR_WAKEUP_PIN2_LOW);
+      }
+    #else
+      HAL_PWR_EnableWakeUpPin(PWR_WAKEUP_PIN2);
+    #endif
+      break;
+#endif //PWR_WAKEUP_PIN2
+#ifdef PWR_WAKEUP_PIN3
+    case 3 :
+    #ifdef STM32L4xx
+      if (mode == RISING) {
+        HAL_PWR_EnableWakeUpPin(PWR_WAKEUP_PIN3_HIGH);
+      }
+      else {
+        HAL_PWR_EnableWakeUpPin(PWR_WAKEUP_PIN3_LOW);
+      }
+    #else
+      HAL_PWR_EnableWakeUpPin(PWR_WAKEUP_PIN3);
+    #endif
+      break;
+#endif //PWR_WAKEUP_PIN3
+#ifdef PWR_WAKEUP_PIN4
+    case 4 :
+    #ifdef STM32L4xx
+      if (mode == RISING) {
+        HAL_PWR_EnableWakeUpPin(PWR_WAKEUP_PIN4_HIGH);
+      }
+      else {
+        HAL_PWR_EnableWakeUpPin(PWR_WAKEUP_PIN4_LOW);
+      }
+    #else
+      HAL_PWR_EnableWakeUpPin(PWR_WAKEUP_PIN4);
+    #endif
+      break;
+#endif //PWR_WAKEUP_PIN4
+#ifdef PWR_WAKEUP_PIN5
+    case 5 :
+    #ifdef STM32L4xx
+      if (mode == RISING) {
+        HAL_PWR_EnableWakeUpPin(PWR_WAKEUP_PIN5_HIGH);
+      }
+      else {
+        HAL_PWR_EnableWakeUpPin(PWR_WAKEUP_PIN5_LOW);
+      }
+    #else
+      HAL_PWR_EnableWakeUpPin(PWR_WAKEUP_PIN5);
+    #endif
+      break;
+#endif //PWR_WAKEUP_PIN5
+    default :
+      break;
+  }
+}
+
+/**
+  * @brief  Enable the sleep mode.
+  * @param  None
+  * @retval None
+  */
+void LowPower_sleep(){
+  /*Suspend Tick increment to prevent wakeup by Systick interrupt.
+    Otherwise the Systick interrupt will wake up the device within 1ms (HAL time base)*/
+  HAL_SuspendTick();
```

Comment:
Shouldn't we disable IRQs and re-enable them after 
using  __disable_irq();   and  __enable_irq();  ?
Otherwise the IRQ handler can be called after HAL_SuspendTick(); and this could have side effects in case the tick is being used by the interrupt handler.

Reply:
same applies to all modes ...

Reply:
If I understood, you suggest something like this:
```
__disable_irq();
HAL_SuspendTick();
__enable_irq();
```
?

Reply:
I was advising __disable_irq before suspending tick and __enable_irq after HAL_ResumeTick

Reply:
Unfortunately you can't because interrupts allow to wake up the MCU. HAL_SuspendTick() disables only Tick IRQ as explained in the comment.

Reply:
Have you tried it? 
I think it should block the SW handling of interrupts (disabled at NVIC level), but it should not block the peripherals interrupts to wake-up the MCU. and yes we still need to stop the TICK interrupt otherwise it would wake-up the MCU.

Reply:
Hope I'm clear enough ... the point is to enter a critical section, not to disable interrupts at peripheral level.